### ISSUE TITLE
Canvas tags unaware of themselves when browser window is inactive

### ISSUE THREAD

**Issue by dovestyle** [goat1000/TagCanvas#8]
I've been having an issue with TagCanvas being unable to bring a "TagToFront" when the browser tab or window is inactive. I've narrowed it down to any tag's "xformed" attribute being undefined.

I use the following code to test:

```javascript
setTimeout(function() { TagCanvas.TagToFront('tag-cloud', {index: 1}); },1000);
```

If I keep the tab with the tag canvas active, it runs fine. However, If I run that code and quickly change tabs, (so the TagToFront call happens while tab is inactive), it breaks with the following message:

```
Uncaught TypeError: Cannot read property 'x' of undefined
TCproto.RotateTag @ jquery.tagcanvas.self.js?body=1:1992
TagCanvas.RotateTag @ jquery.tagcanvas.self.js?body=1:2047
TagCanvas.TagToFront @ jquery.tagcanvas.self.js?body=1:2039
````

The "undefined" object to which the error is referring is the tag's "xformed" attribute. I've even gone as far as console.logging the xformed attribute, and it is indeed undefined when the browser tab is inactive.

Any idea why the tab's active status would cause this behavior?

**Comment by goat1000** [comment]
I haven't been able to reproduce this - are you seeing it in a specific browser or hardware, or everywhere? If you have a test page I could look at, that might help me figure out what is going on.

The `Tag.xformed` variable should only be undefined before the tag has been drawn for the first time, so the timeout must be happening before that. I can't think of a good reason why that would happen though.

**Comment by dovestyle** [comment]
Your comment sparked an idea..

So I use TagCanvas for the appearance of shuffling random numbers. On a 1.5-second interval, I create 30 new random tags and do a Reload. After a set number of seconds - say 30 - I choose another random number, add it to the tags, call Reload again, then call TagToFront to bring the new random number to the front.

In console, if I clearInterval on that 1.5-second random tag generator, and then use the above code, it works just fine on an inactive browser tab. Keep in mind, however, I have not called Reload in console; I just use what's already in the tag list.

When I add a new tag, then call Reload and TagToFront, it does not work in the inactive tab.

What I think this all means is, my call to TagToFront is processing before the Reload is finished drawing the new tag. Maybe Reload also needs a callback ability?

**Comment by goat1000** [comment]
Thanks for the extra info - I've been able to reproduce the bug now. The problem is that `Reload` doesn't actually redraw the tags, it just updates the list. The tags are always drawn in response to a `requestAnimationFrame` event (or `setTimeout` if the browser is ancient) so the tags are not drawn when the page is inactive.

I have a fix for it, I just have to do some testing to make sure it doesn't cause any other problems.

**Comment by goat1000** [comment]
I've committed the fix to master: 29835bff9667c482aa34458c9784cd95915c57c9

Let me know if it works (or not).

**Comment by dovestyle** [comment]
Yep that fixed it! Nice work. Thanks a bunch. Closing issue now.
